### PR TITLE
Add 'veloren'

### PR DESCRIPTION
**Pull request by tim77**
Veloren is a multiplayer voxel RPG written in Rust. It is inspired by games such as Cube World, Legend of Zelda: Breath of the Wild, Dwarf Fortress and Minecraft.

Upstream discussion: https://gitlab.com/veloren/veloren/-/issues/267

### CODE REVIEW

**Comment by tim77** [comment]
bot, build net.veloren.veloren

File: net.veloren.veloren.yaml [hunk 1]
```diff
@@ -0,0 +1,46 @@
+app-id: net.veloren.veloren
+runtime: org.freedesktop.Platform
+runtime-version: '19.08'
+sdk: org.freedesktop.Sdk
+sdk-extensions:
+  - org.freedesktop.Sdk.Extension.rust-nightly
+command: veloren-voxygen
+finish-args:
+  - "--device=all"
+  - "--device=dri"
```
Comment (bilelmoussaoui) [inline net.veloren.veloren.yaml:10]:
```suggestion
```
you're already using `--device=all`

**Comment by tim77** [comment]
Failed because of Git LFS. Git LFS works on localhost with enabled network, but how we can fix it for Flathub build where network not allowed during build?

**Comment by barthalion** [comment]
You need to write a generator (https://github.com/flatpak/flatpak-builder-tools/) and download files before build. Alternatively ask upstream to provide assets in separate archive.

**Comment by tim77** [comment]
> You need to write a generator (https://github.com/flatpak/flatpak-builder-tools/) and download files before build. Alternatively ask upstream to provide assets in separate archive.

They are already downloaded. And this all possible only because of [recent changes ](https://github.com/flatpak/flatpak-builder-tools/commits/master) which @gasinvein made for Cargo generator since Veloren using many own sources.

It turns out that this because `git-lfs` not installed on host system, not because of network during build.

**Comment by barthalion** [comment]
I don't think I follow how recent changes to cargo generator affect downloading LFS assets.

**Comment by gasinvein** [comment]
@barthalion It's unrelated.
As far as I understand, in flatpak-builder initial git checkouts (before module build starts) happen on host, thus, Git LFS support is required on host. And since we have some shrunk files in Flathub build - it looks like Git LFS support is missing on build machine.

**Comment by barthalion** [comment]
It's not enough that `git-lfs` is present, git also needs to be configured to use it. I don't find it sustainable in a wider flatpak context – either there should be a generator or someone interested in contributing support to flatpak-builder needs 1) figure out how LFS affects bare clones 2) make `f-b` pass LFS options to `git clone` which uses the bare repo from step 1.

**Comment by gasinvein** [comment]
@barthalion At least on Fedora installing `git-lfs` package (it runs `git-lfs install --system --skip-repo` via post-install script) is enough to make this work. Can't say for other distros, but I doubt that specific support from flatpak-builder side is required.

**Comment by barthalion** [comment]
And I'm sure it is required. Quick test with bare repo shows that.

```
Cloning into '1234'...
done.
Downloading assets/voxygen/audio/sfx/footsteps/stepgrass_1.wav (47 KB)
Error downloading object: assets/voxygen/audio/sfx/footsteps/stepgrass_1.wav (ace7b9c): Smudge error: Error downloading assets/voxygen/audio/sfx/footsteps/stepgrass_1.wav (ace7b9cf326fdde1c979983fcc549f3864ed2d370c6d5bd89c65e066ff7e4abe): EOF

Errors logged to /tmp/1234/.git/lfs/logs/20200413T132645.651151509.log
Use `git lfs logs last` to view the log.
error: external filter 'git-lfs filter-process' failed
fatal: assets/voxygen/audio/sfx/footsteps/stepgrass_1.wav: smudge filter lfs failed
warning: Clone succeeded, but checkout failed.
You can inspect what was checked out with 'git status'
and retry with 'git restore --source=HEAD :/'
```

**Comment by gasinvein** [comment]
@barthalion Just tried it on a clean VM with Ubuntu 18.04, can confirm that after installing `git-lfs` package and running `sudo git-lfs install --system --skip-repo`, this flatpak can be built as usual, via `flatpak-builder --sandbox`.

Looks like lfs objects are downloaded at `checkout` stage:
```
========================================================================
Building module veloren in /home/gasinvein/.flatpak-builder/build/veloren-1
========================================================================
FB: Running: cp -al /home/gasinvein/.flatpak-builder/git/https_gitlab.com_veloren_veloren /home/gasinvein/.flatpak-builder/build/veloren-1/.git
FB: Running: git config --bool core.bare false
FB: Running: git checkout b751fcc3937d8205cb2261c056aed7ffba218151
Filtering content: 100% (1421/1421), 87.08 MiB | 1.22 MiB/s, done.
Note: checking out 'b751fcc3937d8205cb2261c056aed7ffba218151'.
```

**Comment by barthalion** [comment]
Indeed, no idea why it didn't work manually. We still need `flatpak-builder --download-only` to run `git lfs fetch --all` as we don't want builders to redownload all assets every time a build is run but it's not a blocker.

**Comment by gasinvein** [comment]
@barthalion So what's the next step to make it work?

**Comment by barthalion** [comment]
bot, build net.veloren.veloren

**Comment by barthalion** [comment]
Bad timing. I'll schedule another build later.

**Comment by barthalion** [comment]
bot, build net.veloren.veloren

**Comment by tim77** [comment]
bot, build net.veloren.veloren

**Comment by tim77** [comment]
bot, build net.veloren.veloren

**Comment by tim77** [comment]
Excluded temporary `ARM` arch due to lack of memory. But build for RPi would be useful here since some teachers for example [asking](https://gitlab.com/veloren/veloren/issues/347) about it.

**Comment by barthalion** [comment]
There are two KDE runtimes builds ongoing so that may affect the arm build. Have you managed to build it locally (with or without flatpak-builder)?

**Comment by frzifus** [comment]
is this version somehow outdated? i get this warning when i try to connect to veloren.net.
`[WARN] Server is running 1dd1318a[2020-04-12-20:20], you are running b751fcc3[2020-04-12-22:14], versions might be incompatible!`

File: net.veloren.veloren.yaml [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+      - mkdir -p /app/share/veloren/
+      - cp -ap assets/ /app/share/veloren/
+
+      # Install desktop file, icon and AppData manifest
+      - install -D -m 644 assets/voxygen/net.veloren.veloren.metainfo.xml -t /app/share/metainfo/
+      - install -D -m 644 assets/voxygen/net.veloren.veloren.desktop -t /app/share/applications/
+      - install -D -m 644 assets/voxygen/net.veloren.veloren.png -t /app/share/icons/hicolor/256x256/apps/
+    sources:
+      - type: git
+        commit: b751fcc3937d8205cb2261c056aed7ffba218151
```
Comment (frzifus) [inline net.veloren.veloren.yaml:43]:
```suggestion
        commit: 1dd1318ae01b0f76dee872440d599e2546402380
```

Comment (gasinvein) [reply net.veloren.veloren.yaml:43]: It's a commit from not yet accepted [merged request](https://gitlab.com/veloren/veloren/-/merge_requests/907). Better use commit from master and add the MR as a patch, I guess.

Comment (tim77) [reply net.veloren.veloren.yaml:43]: I can't accept this until [this MR](https://gitlab.com/veloren/veloren/-/merge_requests/907) will be merged.

Comment (tim77) [reply net.veloren.veloren.yaml:43]: Or we can just apply this patch. This is just warning and there is just few commit differs in between. Online should work. Anyway we will wait for `0.6` tag i suppose before final merge. Offline mode works. We can't rebuild for every new commit, at least manually.

Comment (frzifus) [reply net.veloren.veloren.yaml:43]: mh.. ok sounds good. when i try to connect, after the warning, get these messages.
````
[WARN] Server is running 1dd1318a[2020-04-12-20:20], you are running b751fcc3[2020-04-12-22:14], versions might be incompatible!
BINCODE ERROR: Custom("invalid value: integer `13`, expected variant index 0 <= i < 2")
BINCODE ERROR: Custom("invalid value: integer `1`, expected variant index 0 <= i < 1")
BINCODE ERROR: Custom("invalid value: integer `13`, expected variant index 0 <= i < 2")
BINCODE ERROR: Custom("invalid value: integer `13`, expected variant index 0 <= i < 2")
BINCODE ERROR: Custom("invalid value: integer `13`, expected variant index 0 <= i < 2")
BINCODE ERROR: Custom("invalid value: integer `13`, expected variant index 0 <= i < 2")
BINCODE ERROR: Custom("invalid value: integer `13`, expected variant index 0 <= i < 2")
BINCODE ERROR: Custom("invalid value: integer `13`, expected variant index 0 <= i < 2")
BINCODE ERROR: Custom("invalid value: integer `13`, expected variant index 0 <= i < 2")
BINCODE ERROR: Custom("invalid value: integer `13`, expected variant index 0 <= i < 2")
BINCODE ERROR: Custom("invalid value: integer `13`, expected variant index 0 <= i < 2")
BINCODE ERROR: Custom("invalid value: integer `13`, expected variant index 0 <= i < 2")
BINCODE ERROR: Custom("invalid value: integer `13`, expected variant index 0 <= i < 2")
BINCODE ERROR: Custom("invalid value: integer `13`, expected variant index 0 <= i < 2")
BINCODE ERROR: Custom("invalid value: integer `13`, expected variant index 0 <= i < 2")
BINCODE ERROR: Custom("invalid value: integer `13`, expected variant index 0 <= i < 2")
BINCODE ERROR: Custom("invalid value: integer `13`, expected variant index 0 <= i < 2")
[ERROR] [session] Failed to tick the scene: ClientError(Network(Bincode(Custom("invalid value: integer `13`, expected variant index 0 <= i < 2"))))
```

Comment (tim77) [reply net.veloren.veloren.yaml:43]: @frzifus fix coming. Let's automate a little build releases info in appdata and build version from master.

**Comment by tim77** [comment]
bot, build net.veloren.veloren

**Comment by frzifus** [comment]
nice works :+1: thx

**Comment by barthalion** [comment]
@tim77 Who besides you should have write access to the new repository?

**Comment by tim77** [comment]
> @tim77 Who besides you should have write access to the new repository?

After some talks in Discord devs said that they need 1 day to discuss.

**Comment by tim77** [comment]
bot, build net.veloren.veloren

**Comment by tim77** [comment]
This time attempt to built for ARM was successful. But without LTO for ARM.

File: net.veloren.veloren.yaml [hunk 3]
```diff
@@ -6,13 +6,12 @@ sdk-extensions:
...
-  - "--share=network"
-  - "--socket=pulseaudio"
-  - "--socket=wayland"
-  - "--socket=x11"
+  - --device=all
+  - --share=ipc
+  - --share=network
+  - --socket=pulseaudio
+  - --socket=wayland
+  - --socket=x11
```
Comment (barthalion) [inline net.veloren.veloren.yaml:14]: Should be `--socket=fallback-x11` instead of `x11`.

**Comment by barthalion** [comment]
One change needed, other than this LGTM. Any news about whom to add?

**Comment by tim77** [comment]
Please add @Songtronix as well. It's a core Veloren dev.

**Comment by LeandroStanger** [comment]
bot, build net.veloren.veloren

**Comment by bilelmoussaoui** [comment]
The game isn't playable at all under Wayland. The window keep flickering all the time. 
image
Is that something you can reproduce?

**Comment by tim77** [comment]
> The game isn't playable at all under Wayland. The window keep flickering all the time.
> Is that something you can reproduce?

Unfortunately i can't test this myself because of NVIDIA GPU. Upstream bug tracker with [Wayland related issues](https://gitlab.com/veloren/veloren/-/issues?scope=all&utf8=%E2%9C%93&state=all&search=wayland).

**Comment by gasinvein** [comment]
@bilelmoussaoui @tim77 Does the game use x11 by default even on wayland session? If so, `--socket=x11` should be reverted. Otherwise, I believe it should be fixed upstream.

**Comment by bilelmoussaoui** [comment]
The game runs only with `--socket=x11`. The  wayland one should be removed for now; until upstream think their game is wayland ready.

File: net.veloren.veloren.yaml [hunk 4]
```diff
@@ -9,9 +9,9 @@ finish-args:
   - --device=all
   - --share=ipc
   - --share=network
-  - --socket=fallback-x11
   - --socket=pulseaudio
   - --socket=wayland
```
Comment (bilelmoussaoui) [inline net.veloren.veloren.yaml:13]:
```suggestion
```
see my last comment

**Comment by LeandroStanger** [comment]
bot, build net.veloren.veloren

**Pull request closed without merging**